### PR TITLE
Filesystem filtering.

### PR DESCRIPTION
**Pull request by paulp**
Pretty sweet. Try for instance

  ./run.sh ./target /mnt '.*[.]class$'

And you should have a /mnt which contains everything from
target except the classfiles.

### CODE REVIEW

File: run.sh [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env bash
+#
+
+sbt "runMain suffuse.fs.filterfs $*"
```
Comment (EECOLOR) [inline run.sh]: call this file `filterfs` or something similar

File: src/main/scala/fs/fusejna.scala [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+      exec("fusermount", "-u", getMountPoint.getPath)
+  )
+  private def doMount(mountPoint: File, blocking: Boolean): this.type = {
+    addUnmountHook(this)
+    super.mount(mountPoint, blocking)
+    this
+  }
+
+  def mountBg(mountPoint: Path): this.type = doMount(mountPoint.toFile, false)
+  def mountFg(mountPoint: Path): this.type = doMount(mountPoint.toFile, true)
```
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: At the moment it's pretty clear, but I really like it when boolean arguments are named: `blocking = false`

File: src/main/scala/fs/fusejna.scala [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    else
+      exec("fusermount", "-u", getMountPoint.getPath)
+  )
+  private def doMount(mountPoint: File, blocking: Boolean): this.type = {
+    addUnmountHook(this)
+    super.mount(mountPoint, blocking)
+    this
+  }
+
+  def mountBg(mountPoint: Path): this.type = doMount(mountPoint.toFile, false)
```
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: What do `Bg` and `Fg` stand for?

File: src/main/scala/fs/fusejna.scala [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  def mountFg(mountPoint: Path): this.type = doMount(mountPoint.toFile, true)
+
+  def fuseContext()         = super.getFuseContext
+  override def getOptions() = fs.defaultOptions
```
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: No need for the `override` keyword here

File: src/main/scala/fs/fusejna.scala [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+/** This makes it easy to modify or extends the behavior of an existing
+ *  filesystem instance by overriding a small handful of methods.
+ */
+abstract class ForwarderFs extends FuseFs {
```
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala:57]: For now I think this is fine, but I really think we need another abstraction layer for custom file systems (like `FilteredFs`), the methods of `FuseFileSystem` are not the easiest to work with.

File: src/main/scala/fs/fusejna.scala [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+/** Our implementation of the DirFiller interface.
+ */
+final class DirFiller(initial: jList[String]) extends DirectoryFiller {
+  def this() = this(jList())
+
+  private var count = 0
+  private val buf = Vector.newBuilder[Path]
+  private def add(s: String): Unit = { buf += path(s) ; count += 1 }
+
+  locally {
```
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: Is this a new default method? I have never seen it.

File: src/main/scala/fs/fusejna.scala [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  def symlink(path: String, target: String): Int                                               = underlying.symlink(path, target)
+  def truncate(path: String, offset: Long): Int                                                = underlying.truncate(path, offset)
+  def unlink(path: String): Int                                                                = underlying.unlink(path)
+  def utimens(path: String, wrapper: TimeBufferWrapper): Int                                   = underlying.utimens(path, wrapper)
+  def write(path: String, buf: Buf, bufSize: Long, writeOffset: Long, info: FileInfo): Int     = underlying.write(path, buf, bufSize, writeOffset, info)
+}
+
+/** Our implementation of the DirFiller interface.
+ */
+final class DirFiller(initial: jList[String]) extends DirectoryFiller {
```
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: This seems not to be in use, is that correct?

File: src/main/scala/fs/fusejna.scala [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  private var count = 0
+  private val buf = Vector.newBuilder[Path]
+  private def add(s: String): Unit = { buf += path(s) ; count += 1 }
+
+  locally {
+    initial.asScala foreach add
+  }
+  def add(files: jIterable[String]): Boolean = andTrue(files.asScala foreach add)
+  def add(files: String*): Boolean           = andTrue(files foreach add)
+  def result(): Vector[Path]                 = try buf.result finally buf.clear()
```
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: This looks like a dangerous method that can cause a lot of trouble if you are unaware of it's implementation. If you really want this method, please rename it to something like `getResultsAndClear`

File: src/main/scala/fs/package.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  type NodeType          = net.fusejna.types.TypeMode.NodeType
+  type OpenMode          = net.fusejna.StructFuseFileInfo.FileInfoWrapper.OpenMode
+  type StatInfo          = net.fusejna.StructStat.StatWrapper
+  type StatWrapper       = net.fusejna.StructStat.StatWrapper
+  type StatvfsWrapper    = net.fusejna.StructStatvfs.StatvfsWrapper
+  type StructFuseContext = net.fusejna.StructFuseContext
+  type TimeBufferWrapper = net.fusejna.StructTimeBuffer.TimeBufferWrapper
+  type Timespec          = net.fusejna.StructTimespec.ByValue
+  type XattrFiller       = net.fusejna.XattrFiller
+  type XattrListFiller   = net.fusejna.XattrListFiller
```
Comment (EECOLOR) [inline src/main/scala/fs/package.scala]: What is the rationale for making the type aliases?

File: src/main/scala/fs/fusejna.scala (hunk 2 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: I'd never use a boolean argument in the first place if I didn't need it for the underlying interface, but I agree. If only I could fucking [label the calls into java](https://issues.scala-lang.org/browse/SI-5896), which is 98% of where I'm going to need it!

File: src/main/scala/fs/fusejna.scala (hunk 3 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: Background and foreground.

File: src/main/scala/package.scala [hunk 10]
```diff
@@ -4,14 +4,15 @@ import scala.util.{ Success, Failure }
...
-  type Try[+A]           = scala.util.Try[A]
+  type Buf     = java.nio.ByteBuffer
+  type Try[+A] = scala.util.Try[A]
+  type uV      = scala.annotation.unchecked.uncheckedVariance
 
   def Try[A](body: => A): Try[A]        = scala.util.Try[A](body)
   def alreadyExists()                   = -EEXIST
+  def andTrue(x: Unit): Boolean         = true
   def doesNotExist()                    = -ENOENT
+  def doto[A](x: A)(f: A => Unit): A    = { f(x) ; x }
```
Comment (EECOLOR) [inline src/main/scala/package.scala]: haha, I like this one

File: src/main/scala/fs/fusejna.scala (hunk 4 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]:
```
[error] /g/idfs/src/main/scala/fs/idfs.scala:16: class idfs inherits conflicting members:
[error]   method getOptions in class FuseFilesystemAdapterFull of type ()Array[String]  and
[error]   method getOptions in trait FuseFs of type ()Array[String]
[error] (Note: this can be resolved by declaring an override in class idfs.)
[error] class idfs private (from: Path) extends util.FuseFilesystemAdapterFull with FuseFs {
[error]       ^
[error] one error found
```

File: src/main/scala/fs/fusejna.scala (hunk 2 above)
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: ...

Yeah, that...

File: src/main/scala/fs/fusejna.scala (hunk 3 above)
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: Maybe we should use the full names then, but still, background and block feel not related. Maybe it's best if we just call them `mountBlocking` and `mountNonBlocking`

Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: Do we need the non-blocking anyway?

File: src/main/scala/fs/fusejna.scala (hunk 5 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala:57]: No doubt, but it's of paramount importance we not get sidetracked trying to improve the FUSE or fuse-jna interface. I have _reams_ of code written for this purpose and it is easily the biggest reason why this thing isn't a reality yet.

File: src/main/scala/fs/fusejna.scala (hunk 6 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: It's old, it's in predef. It's a no-op. It's to make constructor code stand out more.

File: src/main/scala/fs/fusejna.scala (hunk 4 above)
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: Ahh, right.

If I feel like it I might move the methods from the `FuseFilesystemAdapterFull`

File: src/main/scala/fs/fusejna.scala (hunk 7 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: Possibly true, I'm pulling chunks of code from old projects.

File: src/main/scala/fs/fusejna.scala (hunk 5 above)
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala:57]: Oh, I totally agree. First have (quite) a few examples lying around, otherwise you keep shuffling with the abstractions. It's one of the first lessons I teach my colleagues :-)

File: src/main/scala/fs/fusejna.scala (hunk 6 above)
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: Nice, I wish I knew that one before. Also prevent escaping `val` definitions

File: src/main/scala/fs/package.scala (hunk 9 above)
Comment (paulp) [inline src/main/scala/fs/package.scala]: Not having to import them every time they come up, or playing "babysit the imports" during refactoring. Most of them are in the fuse-jna method signatures. From my POV anytime a name is unambiguous there is little to no downside to aliasing it throughout the package.

Comment (paulp) [inline src/main/scala/fs/package.scala]: I see there's some duplication because I renamed a few on the fly as I pulled this chunk of code in. I'll fix that at least.

**Comment by EECOLOR** [comment]
All in all it looks like a step in the right direction :-)

File: src/main/scala/fs/fusejna.scala (hunk 8 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: I just called it result because the DirectoryFiller is a pure builder and the scala builder interface calls its finishing method result. You aren't allowed to use a builder after you extract its result. It's true this isn't marked as a builder.

Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: In fact after they call result, _I_ can't legally call result on the underlying vectorbuilder again. I could make result a lazy val or something so it can't be called twice.

File: src/main/scala/fs/fusejna.scala (hunk 4 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: Let's do everything possible to leave the fuse-jna code as it is. As soon as we touch it we own a fork. As yet it's a one line source dependency.

File: src/main/scala/fs/fusejna.scala (hunk 3 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala]: Blocking in this context is a synonym for foreground. It means the call to mount won't return until the filesystem is unmounted, whereas in the background version it returns immediately.

If we didn't have the background version we'd have to reimplement it. For instance a command-line suffuse-mount would start everything backgrounded, just like the real mount does.

File: src/main/scala/fs/fusejna.scala (hunk 4 above)
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: You misunderstood. `FuseFileSystemAdapterFull` extends `FuseFileSystem` with default implementations. At one point I will move the remaining default implementations from there into `ForwarderFs`. That would allow us to remove the override, but more importantly, it would remove some of the magic introduced by that adapter.

File: src/main/scala/fs/fusejna.scala (hunk 3 above)
Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala]: Ahh, it's that type of blocking, haha

File: src/main/scala/fs/fusejna.scala (hunk 5 above)
Comment (paulp) [inline src/main/scala/fs/fusejna.scala:57]: I seem to be bent on re-learning the lesson for the Nth time.

Comment (EECOLOR) [inline src/main/scala/fs/fusejna.scala:57]: It's fine to keep you inner 'I want things to feel pretty' person happy, just don't let him get into a state of trance. Especially in exploratory projects like these the questions should constantly be: 'Will this make me go faster next week?' or 'Will this make me go slower next week?'.

Every once in a while you take a few days to refactor the things that have been bugging you.

Comment (paulp) [inline src/main/scala/fs/fusejna.scala:57]: If we can keep it going anything like these last couple days it looks good.
